apps/frontend/src/libs/components/table/table.tsx

```diff
@@ -7,16 +7,28 @@ import {
 import { EMPTY_LENGTH } from "~/libs/constants/constants.js";
 import { type TableColumn } from "~/libs/types/types.js";
 
+import { SelectRowCell } from "./components/components.js";
 import styles from "./styles.module.css";
 
-type Properties<T extends object> = {
+type BaseProperties<T> = {
 	columns: TableColumn<T>[];
 	data: T[];
 };
 
+type SelectableProperties<T> = {
+	getRowId: (row: T) => number;
+	onRowSelect: (rowId: number) => void;
+	selectedRowIds: number[];
+};
+
+type Properties<T> =
+	| BaseProperties<T>
+	| (BaseProperties<T> & SelectableProperties<T>);
+
 const Table = <T extends object>({
 	columns,
 	data,
+	...selectableProperties
```

Comment:
let's get rid of this grouping, it's a bit confusing. We only need this for the types
```suggestion
	getRowId,
	onRowSelect,
	selectedRowIds
```

Reply:
If I do it like that I get errors: Property 'getRowId' does not exist on type 'Properties<T>' ... Cannot make this and [that](https://github.com/BinaryStudioAcademy/bsa-2024-gitfit/pull/176#discussion_r1749321117) work together

Reply:
In this case let's retrieve those property right at the start of the function body:
```
const { getRowId } = selectableProperties
````

Reply:
Now there are errors like
`Property 'getRowId' does not exist on type '{} | { getRowId: (row: T) => number; onRowSelect: (rowId: number) => void; selectedRowIds: number[]; }'.`

A way to fix it is 
```
	const { getRowId, onRowSelect, selectedRowIds } = 
    "getRowId" in selectableProperties && 
    "onRowSelect" in selectableProperties && 
    "selectedRowIds" in selectableProperties 
    ? selectableProperties 
    : { getRowId: undefined, onRowSelect: undefined, selectedRowIds: undefined };

	const isRowSelectable = Boolean(getRowId);
```

But here I need to use getRowId or other selectable prop because isRowSelectable gives a bunch of ts errors
```
<tr className={styles["table-row"]} key={row.id}>
								{getRowId && (
									<td className={styles["table-data"]}>
										<SelectRowCell
											id={getRowId(row.original)}
											isChecked={selectedRowIds.includes(
												getRowId(row.original),
											)}
											onToggle={onRowSelect}
										/>
									</td>
								)}
```

Reply:
oh, I see what's the issue. let's try to use it this way. A small trick to handle ts discriminated unions with destructuring.

```tsx
	const { getRowId, onRowSelect, selectedRowIds } = selectableProperties as
		| SelectableProperties<T>
		| Record<keyof SelectableProperties<T>, undefined>;

	const isRowSelectable = typeof onRowSelect === "function";
```

Reply:
Thanks! That works perfectly. I'll definitely remember this trick.